File: app/src/main/java/com/atrainingtracker/trainingtracker/fragments/aftermath/WorkoutDetailsViewHolder.java
package com.atrainingtracker.trainingtracker.fragments.aftermath;

import android.content.Context;
import android.database.Cursor;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.atrainingtracker.R;
import com.atrainingtracker.trainingtracker.MyHelper;
import com.atrainingtracker.banalservice.BSportType;
import com.atrainingtracker.banalservice.sensor.SensorType;
import com.atrainingtracker.banalservice.sensor.formater.DistanceFormatter;
import com.atrainingtracker.banalservice.sensor.formater.TimeFormatter;
import com.atrainingtracker.banalservice.sensor.formater.SpeedFormatter;
import com.atrainingtracker.banalservice.sensor.formater.PaceFormatter;
import com.atrainingtracker.trainingtracker.database.ExtremaType;
import com.atrainingtracker.trainingtracker.database.WorkoutSummariesDatabaseManager;
import com.atrainingtracker.trainingtracker.database.WorkoutSummariesDatabaseManager.WorkoutSummaries;

import java.util.Locale;

/**
 * A ViewHolder that manages the self-contained 'workout_summary__details.xml' layout.
 */
public class WorkoutDetailsViewHolder {

    private final View mView;
    private final Context mContext;

    // Formatters
    private final DistanceFormatter mDistanceFormatter;
    private final TimeFormatter mTimeFormatter;
    private final SpeedFormatter mSpeedFormatter;
    private final PaceFormatter mPaceFormatter;

    // Views
    private final TextView mTvDistance, mTvMaxDisplacement, mTvTimeActive, mTvTimeTotal;
    private final LinearLayout mRowSpeed, mRowAltitude;
    private final TextView mTvAvgSpeed, mTvAscent, mTvDescent, mTvMinAltitude, mTvMaxAltitude;

    public WorkoutDetailsViewHolder(@NonNull View view, Context context) {
        mView = view;
        mContext = context;

        mDistanceFormatter = new DistanceFormatter();
        mTimeFormatter = new TimeFormatter();
        mSpeedFormatter = new SpeedFormatter();
        mPaceFormatter = new PaceFormatter();

        mTvDistance = view.findViewById(R.id.detail_distance);
        mTvMaxDisplacement = view.findViewById(R.id.detail_max_displacement);
        mTvTimeActive = view.findViewById(R.id.detail_time_active);
        mTvTimeTotal = view.findViewById(R.id.detail_time_total);
        mRowSpeed = view.findViewById(R.id.workout_details_rowSpeed);
        mTvAvgSpeed = view.findViewById(R.id.detail_avg_speed);
        mRowAltitude = view.findViewById(R.id.workout_details_rowAltitude);
        mTvAscent = view.findViewById(R.id.detail_ascent);
        mTvDescent = view.findViewById(R.id.detail_descent);
        mTvMinAltitude = view.findViewById(R.id.detail_min_altitude);
        mTvMaxAltitude = view.findViewById(R.id.detail_max_altitude);
    }

    public View getView() {
        return mView;
    }

    /**
     * Main binding method. Coordinates calls to helper methods to populate the view.
     */
    public void bind(@NonNull Cursor cursor, long workoutId, BSportType bSportType) {
        bindDistanceAndTime(cursor, workoutId);
        bindSpeedAndPace(cursor, bSportType);
        bindAltitude(cursor, workoutId);

        // TODO: Hide the entire container if no data is bound (optional, but good practice)
        mView.setVisibility(View.VISIBLE);
    }

    /**
     * Binds data for the first row: Total Distance, Max Displacement, Active Time, Total Time.
     */
    private void bindDistanceAndTime(@NonNull Cursor cursor, long workoutId) {
        // Total Distance
        double distanceValue = cursor.getDouble(cursor.getColumnIndexOrThrow(WorkoutSummaries.DISTANCE_TOTAL_m));
        String distance = mDistanceFormatter.format(distanceValue);
        String distanceUnit = mContext.getString(MyHelper.getDistanceUnitNameId());
        mTvDistance.setText(mContext.getString(R.string.value_unit_string_string, distance, distanceUnit));

        // Max Displacement
        Double maxDisplacement = WorkoutSummariesDatabaseManager.getExtremaValue(workoutId, SensorType.LINE_DISTANCE_m, ExtremaType.MAX);
        if (maxDisplacement != null) {
            String maxDisplacementFormatted = mDistanceFormatter.format(maxDisplacement);
            String maxDisplacementString = mContext.getString(R.string.value_unit_string_string, maxDisplacementFormatted, distanceUnit);
            mTvMaxDisplacement.setText(mContext.getString(R.string.format_max_displacement, maxDisplacementString));
            mTvMaxDisplacement.setVisibility(View.VISIBLE);
        } else {
            mTvMaxDisplacement.setVisibility(View.GONE);
        }

        // Active Time
        int activeTime = cursor.getInt(cursor.getColumnIndexOrThrow(WorkoutSummaries.TIME_ACTIVE_s));
        mTvTimeActive.setText(mTimeFormatter.format(activeTime));

        // Total Time
        int totalTime = cursor.getInt(cursor.getColumnIndexOrThrow(WorkoutSummaries.TIME_TOTAL_s));
        String totalTimeString = mTimeFormatter.format(totalTime);
        mTvTimeTotal.setText(mContext.getString(R.string.total_time_format, totalTimeString));
    }

    /**
     * Binds data for the speed/pace row, choosing the correct format based on sport type.
     */
    private void bindSpeedAndPace(@NonNull Cursor cursor, BSportType bSportType) {
        float avgSpeed_mps = cursor.getFloat(cursor.getColumnIndexOrThrow(WorkoutSummaries.SPEED_AVERAGE_mps));
        String formattedValue;
        String unit;

        if (bSportType == BSportType.RUN) { // Pace for running
            float pace_spm = (avgSpeed_mps > 0) ? 1 / avgSpeed_mps : 0;
            formattedValue = mPaceFormatter.format(pace_spm);
            unit = mContext.getString(MyHelper.getPaceUnitNameId());
        } else { // Speed for all other sports
            formattedValue = mSpeedFormatter.format(avgSpeed_mps);
            unit = mContext.getString(MyHelper.getSpeedUnitNameId());
        }

        mTvAvgSpeed.setText(mContext.getString(R.string.value_unit_string_string, formattedValue, unit));
        mRowSpeed.setVisibility(View.VISIBLE);
    }

    /**
     * Binds all data for the altitude row: Ascent, Descent, Min/Max Altitude.
     * Manages the visibility of the entire row.
     */
    private void bindAltitude(@NonNull Cursor cursor, long workoutId) {
        boolean hasAltitudeData = false;

        // Ascent
        int ascentM = cursor.getInt(cursor.getColumnIndexOrThrow(WorkoutSummaries.ASCENDING));
        if (ascentM > 0) {
            setTextViewWithIntValue(mTvAscent, ascentM, "m");
            hasAltitudeData = true;
        } else {
            mTvAscent.setVisibility(View.GONE);
        }

        // Descent
        int descentM = cursor.getInt(cursor.getColumnIndexOrThrow(WorkoutSummaries.DESCENDING));
        if (descentM > 0) {
            setTextViewWithIntValue(mTvDescent, descentM, "m");
            hasAltitudeData = true;
        } else {
            mTvDescent.setVisibility(View.GONE);
        }

        // Min Altitude
        Double minAlt = WorkoutSummariesDatabaseManager.getExtremaValue(workoutId, SensorType.ALTITUDE, ExtremaType.MIN);
        if (setTextViewWithDoubleValue(mTvMinAltitude, minAlt, "m")) {
            hasAltitudeData = true;
        }

        // Max Altitude
        Double maxAlt = WorkoutSummariesDatabaseManager.getExtremaValue(workoutId, SensorType.ALTITUDE, ExtremaType.MAX);
        if (setTextViewWithDoubleValue(mTvMaxAltitude, maxAlt, "m")) {
            hasAltitudeData = true;
        }

        mRowAltitude.setVisibility(hasAltitudeData ? View.VISIBLE : View.GONE);
    }

    /**
     * Helper to format an integer value with a unit and set it on a TextView.
     */
    private void setTextViewWithIntValue(@NonNull TextView textView, int value, @NonNull String unit) {
        textView.setText(String.format(Locale.getDefault(), "%d %s", value, unit));
        textView.setVisibility(View.VISIBLE);
    }

    /**
     * Helper to format a Double value with a unit and set it on a TextView.
     * Handles null values by hiding the view.
     * @return True if the view was made visible, false otherwise.
     */
    private boolean setTextViewWithDoubleValue(@NonNull TextView textView, @Nullable Double value, @NonNull String unit) {
        if (value != null) {
            textView.setText(String.format(Locale.getDefault(), "%.0f %s", value, unit));
            textView.setVisibility(View.VISIBLE);
            return true;
        } else {
            textView.setVisibility(View.GONE);
            return false;
        }
    }
}

File: app/src/main/java/com/atrainingtracker/trainingtracker/ui/components/workoutdetails/WorkoutDetailsViewHolder.kt
package com.atrainingtracker.trainingtracker.ui.components.workoutdetails

import android.content.Context
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.isVisible
import com.atrainingtracker.R
import com.atrainingtracker.banalservice.BSportType
import com.atrainingtracker.banalservice.sensor.formater.DistanceFormatter
import com.atrainingtracker.banalservice.sensor.formater.PaceFormatter
import com.atrainingtracker.banalservice.sensor.formater.SpeedFormatter
import com.atrainingtracker.banalservice.sensor.formater.TimeFormatter
import com.atrainingtracker.trainingtracker.MyHelper

/**
 * A "dumb" ViewHolder responsible only for displaying pre-processed WorkoutDetailsData.
 * It has no knowledge of the database or data source.
 * Written in idiomatic Kotlin.
 */
class WorkoutDetailsViewHolder(
    val view: View,
    private val context: Context,
) {

    // Formatters
    private val distanceFormatter = DistanceFormatter()
    private val timeFormatter = TimeFormatter()
    private val speedFormatter = SpeedFormatter()
    private val paceFormatter = PaceFormatter()

    // Views
    private val tvDistance: TextView = view.findViewById(R.id.detail_distance)
    private val tvMaxDisplacement: TextView = view.findViewById(R.id.detail_max_displacement)
    private val tvTimeActive: TextView = view.findViewById(R.id.detail_time_active)
    private val tvTimeTotal: TextView = view.findViewById(R.id.detail_time_total)
    private val rowSpeed: LinearLayout = view.findViewById(R.id.workout_details_rowSpeed)
    private val tvAvgSpeed: TextView = view.findViewById(R.id.detail_avg_speed)
    private val rowAltitude: LinearLayout = view.findViewById(R.id.workout_details_rowAltitude)
    private val tvAscent: TextView = view.findViewById(R.id.detail_ascent)
    private val tvDescent: TextView = view.findViewById(R.id.detail_descent)
    private val tvMinAltitude: TextView = view.findViewById(R.id.detail_min_altitude)
    private val tvMaxAltitude: TextView = view.findViewById(R.id.detail_max_altitude)

    /**
     * Main binding method. Accepts a data object, completely decoupling it from the data source.
     */
    fun bind(data: WorkoutDetailsData) {
        // The ViewHolder becomes a simple coordinator for its private helper methods.
        bindDistanceAndTime(data)
        bindSpeedAndPace(data)
        bindAltitude(data)
        view.isVisible = true
    }

    /**
     * Binds data using the pre-filled data object.
     */
    private fun bindDistanceAndTime(data: WorkoutDetailsData) {
        val distanceUnit = context.getString(MyHelper.getDistanceUnitNameId())

        tvDistance.text = context.getString(
            R.string.value_unit_string_string,
            distanceFormatter.format(data.totalDistance),
            distanceUnit
        )

        data.maxDisplacement?.let {
            val maxDisplacementFormatted = distanceFormatter.format(it)
            val maxDisplacementString = context.getString(
                R.string.value_unit_string_string,
                maxDisplacementFormatted,
                distanceUnit
            )
            tvMaxDisplacement.text = context.getString(R.string.format_max_displacement, maxDisplacementString)
            tvMaxDisplacement.isVisible = true
        } ?: run {
            tvMaxDisplacement.isVisible = false
        }

        tvTimeActive.text = timeFormatter.format(data.activeTimeSec)
        tvTimeTotal.text = context.getString(R.string.total_time_format, timeFormatter.format(data.totalTimeSec))
    }

    /**
     * Binds speed/pace data using the pre-filled data object.
     */
    private fun bindSpeedAndPace(data: WorkoutDetailsData) {
        val formattedValue: String
        val unit: String

        if (data.sportType == BSportType.RUN) {
            val paceSpm = if (data.avgSpeedMps > 0) 1 / data.avgSpeedMps else 0f
            formattedValue = paceFormatter.format(paceSpm.toDouble())
            unit = context.getString(MyHelper.getPaceUnitNameId())
        } else {
            formattedValue = speedFormatter.format(data.avgSpeedMps.toDouble())
            unit = context.getString(MyHelper.getSpeedUnitNameId())
        }

        tvAvgSpeed.text = context.getString(R.string.value_unit_string_string, formattedValue, unit)
        rowSpeed.isVisible = true
    }

    /**
     * Binds altitude data using the pre-filled data object.
     */
    private fun bindAltitude(data: WorkoutDetailsData) {
        var hasAltitudeData = false

        // Using takeIf makes this more concise
        data.ascentMeters.takeIf { it > 0 }?.let {
            tvAscent.setTextViewWithIntValue(it, "m")
            hasAltitudeData = true
        } ?: run { tvAscent.isVisible = false }

        data.descentMeters.takeIf { it > 0 }?.let {
            tvDescent.setTextViewWithIntValue(it, "m")
            hasAltitudeData = true
        } ?: run { tvDescent.isVisible = false }


        if (tvMinAltitude.setTextViewWithDoubleValue(data.minAltitude, "m")) {
            hasAltitudeData = true
        }
        if (tvMaxAltitude.setTextViewWithDoubleValue(data.maxAltitude, "m")) {
            hasAltitudeData = true
        }

        rowAltitude.isVisible = hasAltitudeData
    }

    /**

     * Helper to format an integer value with a unit and set it on a TextView.
     * Defined as an extension function for conciseness.
     */
    private fun TextView.setTextViewWithIntValue(value: Int, unit: String) {
        text = "$value $unit" // String template
        isVisible = true
    }

    /**
     * Helper to format a Double value with a unit and set it on a TextView.
     * Handles null values by hiding the view.
     * @return True if the view was made visible, false otherwise.
     */
    private fun TextView.setTextViewWithDoubleValue(value: Double?, unit: String): Boolean {
        return value?.let {
            text = "%.0f %s".format(it, unit) // String.format
            isVisible = true
            true
        } ?: run {
            isVisible = false
            false
        }
    }
}